File: goals_social_network/lib/models/goal.dart
import 'package:goals_social_network/models/goal_post.dart';
import 'package:goals_social_network/models/user.dart';

class Goal {
  final int id;
  final String title;
  final String description;
  bool done;
  final int status;
  final User userOwner;
  final DateTime updatedAt;
  final List<GoalPost> progressPosts;

  Goal({
    required this.id,
    required this.title,
    required this.description,
    required this.done,
    required this.status,
    required this.userOwner,
    required this.updatedAt,
    required this.progressPosts,
  });

  factory Goal.fromMap(Map goalMap) {
    return Goal(
      id: goalMap['id'],
      title: goalMap['title'],
      description: goalMap['description'],
      done: goalMap['done'],
      status: goalMap['status'],
      updatedAt: DateTime.parse(goalMap['updatedAt']),
      progressPosts: goalMap['progressPosts'] == null
          ? []
          : GoalPost.fromMapList(goalMap['progressPosts']),
      userOwner: User.fromMap(goalMap['userOwner']),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'title': title,
      'description': description,
      'done': done,
    };
  }

  void toggleStatus() {
    done = !done;
  }
}

File: course/src/main/java/goals/social/network/course/models/Goal.java
package goals.social.network.course.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonIncludeProperties;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.*;

@Getter
@Setter
@Entity
@RequiredArgsConstructor
@Table(name = "goals")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Goal {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String description;

    @Column(nullable = false)
    private Boolean done;

    @Column(updatable = false)
    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column()
    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Transient
    private int status;

    public Goal(String title, String description, boolean done, User userOwner) {
        this.title = title;
        this.description = description;
        this.done = done;
        this.userOwner = userOwner;
    }

    public Map<String, Object> toMap() {
        return Map.of(
                "id", id,
                "title", title,
                "description", description,
                "done", done,
                "status", status,
                "userOwner", userOwner.toMap(),
                "updatedAt", updatedAt
        );
    }

    @JsonIncludeProperties(value = {"id", "firstName", "lastName", "email", "phone"})
    @ManyToOne()
    @JoinColumn(name="user_id", referencedColumnName = "id", nullable = false)
    private User userOwner;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "goal")
    @Fetch(FetchMode.SUBSELECT)
    private Set<UserGoalRelations> usersFollowing = new HashSet<>();

    @JsonIncludeProperties(value = {"id", "description", "postReviews", "updatedAt"})
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "goal")
    @Fetch(FetchMode.SUBSELECT)
    private List<GoalProgressPost> progressPosts = new ArrayList<>();

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "goal")
    @Fetch(FetchMode.SUBSELECT)
    private List<GoalInviteCode> inviteCodes = new ArrayList<>();
}
